text/0000-const-trait-impls.md

```diff
@@ -0,0 +1,1045 @@
+}
+```
+
+which now need a `T: (const) Destruct` bound, too.
+In practice we have noticed that a large portion of APIs will have a `(const) Destruct` bound.
+This bound has little value as an explicit bound that appears almost everywhere.
+Especially since it is a fairly straight forward assumption that a type that has trait impls with `const` methods will also have a `Drop::drop` method that is `const` or only contain `const Destruct` types.
+
+In the future we will also want to support `dyn (const) Trait` bounds, which invariably will require the type to implement `(const) Destruct` in order to fill in the function pointer for the `drop` slot in the vtable.
+While that can in generic contexts always be handled by adding more `(const) Destruct` bounds, it would be more similar to how normal `dyn` safety
+works if there were implicit `(const) Destruct` bounds for (most?) `(const) Trait` bounds.
+
+Thus we require that all `trait`s with `(const)` methods also have a `(const) Destruct` super trait bound to ensure users don't need to add `(const) Destruct` bounds everywhere.
+We may relax this requirement in the future or make it implied.
+
+## `(const)` bounds on `Drop` impls
+
+It is legal to add `(const)` to `Drop` impls' bounds, even thought the struct doesn't have them:
+
+```rust
+trait Bar {
+    (const) fn thing(&mut self);
+}
+
+struct Foo<T: Bar>(T);
+
+impl<T: (const) Bar> Drop for Foo<T> {
+    const fn drop(&mut self) {
+        self.0.thing();
+    }
+}
+```
+
+There is no reason (and no coherent representation) of adding `(const)` trait bounds to a type.
+Our usual `Drop` rules enforce that an impl must have the same bounds as the type.
+`(const)` modifiers are special here, because they are only needed in const contexts.
+While they cause exactly the divergence that we want to prevent with the `Drop` impl rules:
+a type can be declared, but not dropped, because bounds are unfulfilled, this is:
+
+* Already the case in const contexts, just for all types that aren't trivially free of `Drop` types.
+* Exactly the behaviour we want.
+
+Extraneous `(const) Trait` bounds where `Trait` isn't a bound on the type at all are still rejected:
+
+```rust
+impl<T: (const) Bar + (const) Baz> Drop for Foo<T> {
+    const fn drop(&mut self) {
+        self.0.thing();
+    }
+}
+```
+
+errors with
+
+```
+error[E0367]: `Drop` impl requires `T: Baz` but the struct it is implemented for does not
+  --> src/lib.rs:13:22
+   |
+13 | impl<T: (const) Bar + (const) Baz> Drop for Foo<T> {
+   |                      ^^^^^^^^^^
+   |
+note: the implementor must specify the same requirement
+  --> src/lib.rs:8:1
+   |
+8  | struct Foo<T: Bar>(T);
+   | ^^^^^^^^^^^^^^^^^^
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+## Adding any feature at all around constness
+
+I think we've reached the point where all critics have agreed that this one kind of effect system is unavoidable since we want to be able to write maintainable code for compile time evaluation.
```

Comment:
```suggestion
I think we've reached the point where all critics have agreed that this one kind of effect system is unavoidable since we want to be able to write maintainable generic code for compile time evaluation.
```